assets/js/collaborative-editor/components/run-viewer/LogTabPanel.tsx

```diff
@@ -19,12 +58,29 @@ export function LogTabPanel() {
     null
   );
   const mountedRef = useRef(false);
+  const prevRunIdRef = useRef<string | null>(null);
 
   // Track log level state from store
   const [logLevel, setLogLevel] = useState<'debug' | 'info' | 'warn' | 'error'>(
     () => storeRef.current.getState().desiredLogLevel as any
   );
 
+  // Track whether we have received any logs
+  const [hasLogs, setHasLogs] = useState(false);
+
+  // Typewriter animation state - rendered as React overlay, NOT in Monaco
+  // Initial message types out first, then stays visible while cycling messages animate below
+  const [initialText, setInitialText] = useState(''); // The typed-out initial message
+  const [cyclingText, setCyclingText] = useState('');
+  const animationRef = useRef<ReturnType<typeof setTimeout> | null>(null);
+  const currentMessageRef = useRef<string>('');
+  const initialTextRef = useRef<string>(''); // Mirror of initialText
+  const cyclingTextRef = useRef<string>(''); // Mirror of cyclingText
+  const animationPhaseRef = useRef<
+    'typing-initial' | 'typing' | 'paused' | 'deleting' | 'waiting'
+  >('typing-initial');
+  const runStateRef = useRef<string | undefined>(undefined);
```

Comment:
These useRefs scare me.

we use them as an escape hatch so we can update some state value without causing a re-render. Looking at `cyclingText` and `cyclingTextRef`. I see that whenever `cyclingText` is set using `setCyclingText` we also update the value in the ref at `cyclingTextRef`. which tells me one is redundant. Same with `initialTextRef`.